File: ios/Domain/Sources/VellumDomain/SessionEngine.swift
import Foundation

public enum SessionEngine {
    public static func createRetrievalEvent(
        id: String,
        concept: Concept,
        sessionId: String,
        promptId: String,
        responseText: String,
        outcome: RetrievalOutcome,
        createdAt: Date
    ) -> LearningEvent {
        let masteryAfter = LearnerModel.applyRetrievalMastery(
            mastery: concept.mastery,
            outcome: outcome,
            difficulty: concept.difficulty,
            successfulRetrievals: concept.successfulRetrievals
        )
        return LearningEvent(
            id: id,
            conceptId: concept.id,
            sessionId: sessionId,
            kind: .retrieval,
            outcome: outcome,
            promptId: promptId,
            responseText: responseText,
            masteryBefore: concept.mastery,
            masteryAfter: masteryAfter,
            createdAt: createdAt
        )
    }

    public static func buildQueue(
        plan: StudyPlan,
        concepts: [Concept],
        prompts: [RetrievalPrompt],
        questionCount: Int = 6
    ) -> [SessionItem] {
        let conceptBlocks = plan.blocks.filter { $0.kind == .concept }
        var items: [SessionItem] = []
        var promptCursor: [String: Int] = [:]
        var index = 0

        func nextPrompt(for conceptId: String) -> RetrievalPrompt? {
            let matches = prompts.filter { $0.conceptId == conceptId }
            guard !matches.isEmpty else { return nil }
            let cursor = promptCursor[conceptId, default: 0]
            promptCursor[conceptId] = cursor + 1
            return matches[cursor % matches.count]
        }

        while items.count < questionCount, !conceptBlocks.isEmpty {
            let block = conceptBlocks[index % conceptBlocks.count]
            index += 1
            guard let conceptId = block.conceptId,
                  let concept = concepts.first(where: { $0.id == conceptId }),
                  let prompt = nextPrompt(for: conceptId)
            else { continue }
            items.append(
                SessionItem(
                    id: "q-\(items.count + 1)-\(prompt.id)",
                    conceptId: concept.id,
                    conceptName: concept.name,
                    promptId: prompt.id,
                    question: prompt.question,
                    modelAnswer: prompt.modelAnswer
                )
            )
            if items.count > 40 { break }
        }
        return items
    }

    public static func summarize(
        plannedMinutes: Int,
        before: [Concept],
        after: [Concept],
        exam: Exam,
        relationships: [ConceptRelationship],
        now: Date
    ) -> SessionSummary {
        let beforeById = Dictionary(uniqueKeysWithValues: before.map { ($0.id, $0) })
        let deltas = after.compactMap { concept -> MasteryDelta? in
            guard let previous = beforeById[concept.id] else { return nil }
            guard abs(concept.mastery - previous.mastery) >= 0.005 else { return nil }
            return MasteryDelta(
                conceptId: concept.id,
                name: concept.name,
                before: previous.mastery,
                after: concept.mastery
            )
        }
        .sorted { abs($0.after - $0.before) > abs($1.after - $1.before) }

        let tomorrow = StudyScheduler.planStudyBlock(
            concepts: after,
            exam: exam,
            relationships: relationships,
            availableMinutes: plannedMinutes,
            now: now.addingTimeInterval(LearnerConstants.secondsPerDay)
        )
        .blocks
        .filter { $0.kind == .concept }
        .prefix(3)
        .map(\.title)

        return SessionSummary(
            plannedMinutes: plannedMinutes,
            readinessBefore: StudyScheduler.readiness(concepts: before),
            readinessAfter: StudyScheduler.readiness(concepts: after),
            conceptDeltas: deltas,
            tomorrowNames: Array(tomorrow)
        )
    }
}

File: domain/session-engine.ts
import { applyRetrievalMastery } from "./learner-model";
import { compareRoutes, generateRoute } from "./routing-engine";
import type { Concept, LearnerSnapshot, LearningEvent, RetrievalOutcome, RouteChange, StudySession } from "./types";

export function masteryAfterEvidence(concept: Concept, outcome: RetrievalOutcome) {
  return applyRetrievalMastery(concept.mastery, outcome, concept.difficulty, concept.successfulRetrievals);
}

export function recalculateSessionRoute(input: {
  snapshot: LearnerSnapshot;
  session: StudySession;
  previousRoute: StudySession["latestRoute"];
  nowIso: string;
}) {
  const exam = input.snapshot.exams.find((item) => item.id === input.session.examId);
  if (!exam) return { route: input.previousRoute, change: null as RouteChange | null };
  const route = generateRoute({
    concepts: input.snapshot.concepts.filter((concept) => concept.courseId === input.session.courseId),
    relationships: input.snapshot.relationships,
    events: input.snapshot.events,
    exam,
    nowIso: input.nowIso,
    availableMinutes: input.previousRoute.availableMinutes,
  });
  return { route, change: compareRoutes(input.previousRoute, route) };
}

export function evidenceEvent(input: {
  id: string;
  userId: string;
  concept: Concept;
  sessionId: string | null;
  promptId: string | null;
  responseText: string | null;
  outcome: RetrievalOutcome;
  responseTimeMs: number | null;
  answerRevealed: boolean;
  createdAt: string;
}): LearningEvent {
  return {
    id: input.id,
    userId: input.userId,
    conceptId: input.concept.id,
    sessionId: input.sessionId,
    kind: "retrieval",
    outcome: input.outcome,
    selfRating: null,
    assistance: input.answerRevealed ? "answer_revealed" : "none",
    responseTimeMs: input.responseTimeMs,
    promptId: input.promptId,
    responseText: input.responseText,
    masteryBefore: input.concept.mastery,
    masteryAfter: masteryAfterEvidence(input.concept, input.outcome),
    createdAt: input.createdAt,
  };
}
